machine/v1/types_nutanixprovider.go

```diff
@@ -154,6 +166,161 @@ type NutanixResourceIdentifier struct {
+	// +optional
+	// +unionMember
+	UUID *string `json:"uuid,omitempty"`
+
+	// name is the name of the storage_container object. It cannot be empty if the type is Name.
+	// +optional
+	// +unionMember
+	Name *string `json:"name,omitempty"`
+
+	// url is the source URL of the storage_container object. It cannot be empty if the type is URL.
+	// +optional
+	// +unionMember
+	URL *string `json:"url,omitempty"`
+}
+
+// NutanixVMStorageConfig specifies the storage configuration parameters for VM disks.
+type NutanixVMStorageConfig struct {
+	// flashMode specifies whether to pin the VM disk to the flash tier.
+	// +optional
+	FlashMode bool `json:"flashMode,omitempty"`
+
+	// storageContainer refers to the storage_container used by the VM disk.
+	// +optional
+	StorageContainer *NutanixStorageContainerReference `json:"storageContainer,omitempty"`
+}
+
+// NutanixDiskAdapterType is the disk device adapter type
+type NutanixDiskAdapterType string
+
+const (
+	// NutanixDiskAdapterTypeSCSI represents the disk adapter type "SCSI".
+	NutanixDiskAdapterTypeSCSI NutanixDiskAdapterType = "SCSI"
+
+	// NutanixDiskAdapterTypeIDE represents the disk adapter type "IDE".
+	NutanixDiskAdapterTypeIDE NutanixDiskAdapterType = "IDE"
+
+	// NutanixDiskAdapterTypePCI represents the disk adapter type "PCI".
+	NutanixDiskAdapterTypePCI NutanixDiskAdapterType = "PCI"
+
+	// NutanixDiskAdapterTypeSATA represents the disk adapter type "SATA".
+	NutanixDiskAdapterTypeSATA NutanixDiskAdapterType = "SATA"
+)
+
+// NutanixDiskAddress specifies the disk address.
+type NutanixDiskAddress struct {
+	// adapterType is the adapter type of the disk address.
+	// +kubebuilder:default=SCSI
+	// +kubebuilder:validation:Enum=SCSI;IDE;PCI;SATA
+	AdapterType NutanixDiskAdapterType `json:"adapterType,omitempty"`
+
+	// deviceIndex is the index of the disk address.
+	DeviceIndex int64 `json:"deviceIndex,omitempty"`
+}
+
+// NutanixDiskDeviceType is the VM disk device type
+type NutanixDiskDeviceType string
+
+const (
+	// NutanixDiskDeviceTypeDisk specifies the VM disk device type is DISK.
+	NutanixDiskDeviceTypeDisk NutanixDiskDeviceType = "DISK"
+
+	// NutanixDiskDeviceTypeCDROM specifies the VM disk device type is DISK.
+	NutanixDiskDeviceTypeCDROM NutanixDiskDeviceType = "CDROM"
+)
+
+// NutanixVMDiskDeviceProperties specifies the
+type NutanixVMDiskDeviceProperties struct {
+	// deviceType is a disk type. The default is DISK.
+	// +kubebuilder:default=DISK
+	// +kubebuilder:validation:Enum=DISK;CDROM
+	DeviceType NutanixDiskDeviceType `json:"deviceType"`
+
+	// diskAddress is the address of disk to boot from.
+	DiskAddress NutanixDiskAddress `json:"diskAddress,omitempty"`
+}
+
+// NutanixDataDisk specifies the VM data disk configuration parameters.
+type NutanixVMDisk struct {
+	// diskSize is size (in Quantity format) of the disk attached to the VM
+	// The minimum diskSize is 1Gi bytes
+	// +kubebuilder:validation:Required
+	DiskSize resource.Quantity `json:"diskSize"`
```

Comment:
Can this disk size be only whole gigabytes in the nutanix APIs or does it resolve down to some lower, eg bytes?

Reply:
diskSize can use any valid Quantity format, not necessary in whole gigabytes. The minimum of 1Gi validation is added to the Machine webhook and the nutanix-mapi-controoler.